Adds more features back that were lost due to the removal of Newtonsoft

gvkries opened the pull request:
Added `ToObject<T>` implementations back to `JsonDynamicObject`, `JsonDynamicArray` and `JsonDynamicValue`. 
This functionality allows to convert JSON to the strongly typed content items and parts when using dynamic typing. E.g. we have this functionality referenced in the documentation:
https://github.com/OrchardCMS/OrchardCore/blob/7196b5b1c6c0180b17bb4dabe944232001b0a681/src/docs/reference/modules/Media/README.md?plain=1#L456

Also something like the following is now possible again:

```csharp
@foreach (var blogPost in Model.ContentItem.Content.NamedBagPart.ContentItems)
{
    var contentItem = blogPost.ToObject<ContentItem>();
    
    [...]
}
```

To support value comparisons when using dynamic typing, I made the following changes:

1. Implemented equality operators in JsonDynamicValue.
2. Added implicit conversions from base types to JsonDynamicValue.
3. Implemented IComparable<T> in JsonDynamicValue.

E.g. this allows using conditions with dynamic values, without casting to the target type before:
```csharp
dynamic blogPost = ...;
if (blogPost.BlogPost.Subtitle.Text == "Test") {... }
```

Note that the comparison implementation (`<, >, <=, >=`) is very basic at the moment, e.g. it does not handle all types
a `JsonDynamicValue` can be converted too. The comparison was therefore supporting a wider range
of target types with Newtonsoft JSON, in contrast to this rudimentary implementation.

Fixes #16233
Fixes #16221
~~Fixes #16290~~

/cc @giannik @MikeAlhayek @sebastienros

hyzx86 commented:
The unit test I mentioned in this issue doesn't work directly here, any plans to fix it in this PR?

https://github.com/OrchardCMS/OrchardCore/issues/16290

Can we modify the DynamicJsonValue fetching behaviour so that it can be generated from the `JsonValueKind` enumeration, e.g. 

JsonValueKind.String map to System.
JsonValueKind.Number map  to Decimal.
JsonValueKind.Object map to DynamicJsonValue.
...others

I'm working on that.

hyzx86 commented:
Similar functionality exists in Jint and ClearScript, Jint uses ExpandoObject, but ClearScript uses DynamicObject. 
Shall we try ExpandoObject as well? I've tried using ClearScript before because of the same type conversion or value compatibility issues.

https://github.com/microsoft/ClearScript/discussions/485

sebastienros commented on src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicValue.cs:
```diff
@@ -363,6 +463,329 @@ ulong IConvertible.ToUInt64(IFormatProvider? provider)
+            return result;
+        }
+
+        // converting a fractional number to a BigInteger will lose the fraction
+        // check for fraction if result is two numbers are equal
```
@MikeAlhayek there is no `long` here

sebastienros commented:
Dude you are so cool, but even more if you create a unit test with these conversions ;)

hishamco commented on src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicObject.cs:
```diff
@@ -39,6 +39,8 @@ public void Merge(JsonNode? content, JsonMergeSettings? settings = null)
         set => SetValue(key, value);
     }
 
+    public T? ToObject<T>() => _jsonObject.ToObject<T>();
```
Can we create a base class and move this to it?

gvkries replied:
Probably, but it doesn't buy us much for this single method.

gvkries replied:
Done.

hishamco commented on src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicValue.cs:
```diff
@@ -24,6 +26,81 @@ public override DynamicMetaObject GetMetaObject(Expression parameter)
+
+        object? otherValue;
+        JsonValueKind valueKind;
+
```
```suggestion
```

hishamco commented on src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicValue.cs:
```diff
@@ -24,6 +26,81 @@ public override DynamicMetaObject GetMetaObject(Expression parameter)
+        return obj.Equals(objOther);
+    }
+
+    public override int GetHashCode()
+    {
+        return JsonValue.GetObjectValue()?.GetHashCode() ?? 0;
+    }
```
```suggestion
    public override int GetHashCode() => JsonValue.GetObjectValue()?.GetHashCode() ?? 0;
```

hishamco commented on src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicValue.cs:
```diff
@@ -148,6 +225,41 @@ ulong IConvertible.ToUInt64(IFormatProvider? provider)
+        return left.Equals(right);
+    }
+
+    public static bool operator !=(JsonDynamicValue left, JsonDynamicValue right)
+    {
+        return !(left == right);
+    }
```
```suggestion
    public static bool operator !=(JsonDynamicValue left, JsonDynamicValue right) => !(left == right);
```

gvkries replied:
I know you prefer expression bodies, but Sebastien removed all of them the last time this file was changed. So I rather keep the format consistent for now.

sebastienros replied:
I confirm. When I tried to open the classes and read what they were doing it was so condensed I couldn't understand what I was reading.

What I realized is that if the definition + body fits on a line on the screen then it's fine, if the implementation has to go on the second line then there is no point and it's more readable to just use a standard body.

sebastienros replied:
But please don't change anything because "that one char can be removed and it will fit in a single line", there is no readability advantage IMO by squashing everything in a line.

hishamco commented on src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicValue.cs:
```diff
@@ -148,6 +225,41 @@ ulong IConvertible.ToUInt64(IFormatProvider? provider)
+        return !(left == right);
+    }
+
+    public static bool operator <(JsonDynamicValue left, JsonDynamicValue right)
+    {
+        return ReferenceEquals(left, null) ? !ReferenceEquals(right, null) : left.CompareTo(right) < 0;
+    }
```
```suggestion
    public static bool operator <(JsonDynamicValue left, JsonDynamicValue right)
        => ReferenceEquals(left, null) ? !ReferenceEquals(right, null) : left.CompareTo(right) < 0;
```

hishamco commented on src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicValue.cs:
```diff
@@ -148,6 +225,41 @@ ulong IConvertible.ToUInt64(IFormatProvider? provider)
+        return ReferenceEquals(left, null) ? !ReferenceEquals(right, null) : left.CompareTo(right) < 0;
+    }
+
+    public static bool operator <=(JsonDynamicValue left, JsonDynamicValue right)
+    {
+        return ReferenceEquals(left, null) || left.CompareTo(right) <= 0;
+    }
+
+    public static bool operator >(JsonDynamicValue left, JsonDynamicValue right)
+    {
+        return !ReferenceEquals(left, null) && left.CompareTo(right) > 0;
+    }
+
+    public static bool operator >=(JsonDynamicValue left, JsonDynamicValue right)
+    {
+        return ReferenceEquals(left, null) ? ReferenceEquals(right, null) : left.CompareTo(right) >= 0;
+    }
```
```suggestion
    public static bool operator <=(JsonDynamicValue left, JsonDynamicValue right)
        => ReferenceEquals(left, null) || left.CompareTo(right) <= 0;

    public static bool operator >(JsonDynamicValue left, JsonDynamicValue right)
        => !ReferenceEquals(left, null) && left.CompareTo(right) > 0;

    public static bool operator >=(JsonDynamicValue left, JsonDynamicValue right)
        => ReferenceEquals(left, null) ? ReferenceEquals(right, null) : left.CompareTo(right) >= 0;
```

hishamco commented on src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicValue.cs:
```diff
@@ -363,6 +475,330 @@ ulong IConvertible.ToUInt64(IFormatProvider? provider)
+
+    public static implicit operator JsonDynamicValue(uint value)
+    {
+        return new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));
+    }
+
+    public static implicit operator JsonDynamicValue(uint? value)
+    {
+        return new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));
+    }
+
+    public static implicit operator JsonDynamicValue(ulong value)
+    {
+        return new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));
+    }
+
+    public static implicit operator JsonDynamicValue(ulong? value)
+    {
+        return new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));
+    }
+
+    public static implicit operator JsonDynamicValue(byte[] value)
+    {
+        return new JsonDynamicValue(JsonValue.Create(Convert.ToBase64String(value), JOptions.Node));
+    }
+
+    public static implicit operator JsonDynamicValue(TimeSpan value)
+    {
+        return new JsonDynamicValue(JsonValue.Create(value.ToString(), JOptions.Node));
+    }
+
+    public static implicit operator JsonDynamicValue(TimeSpan? value)
+    {
+        return new JsonDynamicValue(value.HasValue ? JsonValue.Create(value.ToString(), JOptions.Node) : null);
+    }
+
+    public static implicit operator JsonDynamicValue(Uri? value)
+    {
+        return new JsonDynamicValue(value != null ? JsonValue.Create(value.ToString(), JOptions.Node) : null);
+    }
```
```suggestion
    public static implicit operator JsonDynamicValue(bool value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(bool? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(byte value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(byte? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(char value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(char? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(DateTime value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(DateTime? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(DateTimeOffset value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(DateTimeOffset? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(decimal value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(decimal? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(double value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(double? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(Guid value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(Guid? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(short value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(short? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(int value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(int? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(long value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(long? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(sbyte value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(sbyte? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(float value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(float? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(string value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(ushort value)
       => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(ushort? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(uint value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(uint? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(ulong value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(ulong? value)
        => new JsonDynamicValue(JsonValue.Create(value, JOptions.Node));

    public static implicit operator JsonDynamicValue(byte[] value)
        => new JsonDynamicValue(JsonValue.Create(Convert.ToBase64String(value), JOptions.Node));

    public static implicit operator JsonDynamicValue(TimeSpan value)
        => new JsonDynamicValue(JsonValue.Create(value.ToString(), JOptions.Node));

    public static implicit operator JsonDynamicValue(TimeSpan? value)
        => new JsonDynamicValue(value.HasValue ? JsonValue.Create(value.ToString(), JOptions.Node) : null);

    public static implicit operator JsonDynamicValue(Uri? value)
        => new JsonDynamicValue(value != null ? JsonValue.Create(value.ToString(), JOptions.Node) : null);
```

gvkries commented:
I added some basic unit tests for the new implicit conversions and operators.

Also, to make it even more complex, I added a fix for #16290 into this PR. Serializing the dynamic wrappers (`JsonDynamicObject`, `JsonDynamicArray` and `JsonDynamicValue`) should not include these objects in the output at all. They should be transparent for users, the same way as it was with NSJ, where we didn't had this extra abstraction layer.

hyzx86 commented:
I've tested it and found that it's still a little different from the old version, but it's almost complete. Thanks!

```js
const content = contentItem('46t41hp7d9w6446r35405swzwt')

return {
    data: {
        // contentStr:JSON.stringify(content), //Error: Cyclic reference detected
        testValue: content.Content.SEAQuotaionForm.SoldTo.Text,
        testValue2: content.Content.SEAQuotaionForm.SoldTo.Text + "2123",
        QuoteId: content.Content.SEAQuotaionForm.QuoteId.Value + 10000,
        obj: JSON.parse(JSON.stringify(content.Content)),
        objStr: JSON.stringify(content.Content),
        objAll: content
    }
}
```


this PR
![image](https://github.com/OrchardCMS/OrchardCore/assets/15613121/a9ae61db-cb26-4b59-a8b9-5d22f0796576)

this is oc 1.8
![image](https://github.com/OrchardCMS/OrchardCore/assets/15613121/8659ccdb-8102-4a7d-a926-ad59f9d3207e)

gvkries commented:
I think in Jint you have the same problem. The dynamic wrappers are now also serialized, that's why you get a different result. The `JsonDynamicObject` only has the `Count` property public, and that's getting serialized.

hyzx86 commented:
> I think in Jint you have the same problem. The dynamic wrappers are now also serialized, that's why you get a different result. The `JsonDynamicObject` only has the `Count` property public, and that's getting serialized. 

Yes, in my tests the serialisation converter you defined works in Jint

hyzx86 commented:
I created a PR for solving serialisation problems in Jint https://github.com/OrchardCMS/OrchardCore/pull/16298

gvkries commented:
I simplified this a little bit by adding a base class for the dynamic wrappers (as @hishamco  suggested). Also, I removed the nested  `JsonDynamicMetaObject` class as it is not required.

hyzx86 commented on src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicBase.cs:
```diff
@@ -0,0 +1,13 @@
+
+public abstract class JsonDynamicBase : DynamicObject
+{
+    public abstract JsonNode? Node { get; }
```
Why `Node` and not `Value`, would you use the member name `Value` to determine the target type?

Can we use something like this to check the target type in this PR?

```C#
 case JsonValueKind.String:
                if (memberName == "Value")
                {
                    if (currentNodeValue.TryGetValue<DateTime>(out var datetime))
                    {
                        result = datetime;
                        return true;
                    }

                    if (currentNodeValue.TryGetValue<TimeSpan>(out var timeSpan))
                    {
                        result = timeSpan;
                        return true;
                    }
                }

                result = currentNodeValue.GetString();
                return true;
```

gvkries replied:
It would be a `Value` property on `JsonDynamicValue` only, not on the base class. This is the base class for `JsonDynamicObject` and `JsonDynamicArray` too, so `Value` would be misleading. 

But getting the target type like this is always only a guess, I'm not sure if we should add that back.

hyzx86 replied:
ok ，I've tested it. It's also read as a string in 1.8, no problem. 
```js
const content = contentItem("4s7hn4k9rf0yfzez8ds7mrs9d2");
return {
    data:{
        year:new Date(content.Content.Test.DateField.Value).getFullYear()
    }
}
```

hyzx86 reviewed (approved):
Thanks !

sebastienros commented on src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicObject.cs:
```diff
@@ -3,17 +3,18 @@
 [DebuggerDisplay("JsonDynamicObject[{Count}]")]
-public class JsonDynamicObject : DynamicObject
+[JsonConverter(typeof(JsonDynamicJsonConverter<JsonDynamicObject>))]
+public class JsonDynamicObject : JsonDynamicBase
```
Can you make all these classes sealed (for perf)? Or do we expect users to be able to sub-class it? If you have some hesitation just use sealed, and we can unseal it later if someone comes with a real use-case.

gvkries replied:
Fair point, I don't think they should be able to sub-class. I'll seal them all.

gvkries replied:
Done.

sebastienros commented:
@MikeAlhayek @hishamco waiting for you asking for more things, or we'll merge in 24h.

MikeAlhayek commented on test/OrchardCore.Tests/Data/JsonDynamicTests.cs:
```diff
@@ -337,4 +341,479 @@ public void JsonDynamicValueMustConvertToUri()
+        Assert.Equal("{\"stringValue\":\"test\",\"numberValue\":123,\"booleanValue\":true}", jsonStr);
+    }
+
+    public class TestPart : ContentPart
```
```suggestion
    public sealed class TestPart : ContentPart
```

MikeAlhayek reviewed (approved):
LGTM

sebastienros commented on test/OrchardCore.Tests/Data/JsonDynamicTests.cs:
```diff
@@ -810,7 +810,7 @@ public void SerializingJsonDynamicValueMustWriteValueOnly()
     }
 
-    public class TestPart : ContentPart
+    public sealed class TestPart : ContentPart
```
sealed is only useful for perf when it has methods overrides, or to prevent users from inheriting it. Here there is no reason to make these changes. The same way there is no reason to remove it now it's there.

hishamco commented:
@sebastienros if there's any objection regarding expression bodies please let me, because it was a useful feature for single-line methods

hyzx86 commented:
> @sebastienros if there's any objection regarding expression bodies please let me, because it was a useful feature for single-line methods

Personally, it's not a big deal, and I'm more than inclined to prioritise readability, too
The more urgent thing right now should be to resolve the remaining almost 1000 issues 😅
If nothing else is added, I suggest merging it as soon as possible

hishamco commented on src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicObject.cs:
```diff
@@ -28,17 +29,19 @@ public JsonDynamicObject(JsonObject jsonObject)
         set => SetValue(key, value);
     }
 
+    public void Merge(JsonNode? content, JsonMergeSettings? settings = null)
+    {
+        _jsonObject.Merge(content, settings);
+    }
```
```suggestion
    public void Merge(JsonNode? content, JsonMergeSettings? settings = null) => _jsonObject.Merge(content, settings);
```

hishamco commented on src/OrchardCore/OrchardCore.Abstractions/Json/Serialization/JsonDynamicJsonConverter.cs:
```diff
@@ -0,0 +1,25 @@
+
+public sealed class JsonDynamicJsonConverter<T> : JsonConverter<T> where T : JsonDynamicBase
+{
+    public override T? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+    {
+        throw new NotSupportedException($"Deserializing a {typeof(T).Name} is not supported.");
+    }
```
```suggestion
    public override T? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        => throw new NotSupportedException($"Deserializing a {typeof(T).Name} is not supported.");
```

hishamco commented on test/OrchardCore.Tests/Data/JsonDynamicTests.cs:
```diff
@@ -337,4 +341,479 @@ public void JsonDynamicValueMustConvertToUri()
+
+    public sealed class TestPart : ContentPart
+    {
+        public TextField TextFieldProp { get; set; }
+        public NumericField NumericFieldProp { get; set; }
+        public BooleanField BooleanFieldProp { get; set; }
```
```suggestion
        public TextField TextFieldProp { get; set; }

        public NumericField NumericFieldProp { get; set; }

        public BooleanField BooleanFieldProp { get; set; }
```

hishamco commented on test/OrchardCore.Tests/Data/JsonDynamicTests.cs:
```diff
@@ -337,4 +341,479 @@ public void JsonDynamicValueMustConvertToUri()
+    [Fact]
+    public void SerializingJsonDynamicValueMustWriteValueOnly()
+    {
+        var contentItem = new ContentItem();
+        contentItem.Alter<TestPart>(part =>
+        {
+            part.TextFieldProp = new TextField { Text = "test" };
+            part.NumericFieldProp = new NumericField { Value = 123 };
+            part.BooleanFieldProp = new BooleanField { Value = true };
+        });
+
+        dynamic expandoValue = new ExpandoObject();
+        expandoValue.stringValue = contentItem.Content.TestPart.TextFieldProp.Text;
+        expandoValue.numberValue = contentItem.Content.TestPart.NumericFieldProp.Value;
+        expandoValue.booleanValue = contentItem.Content.TestPart.BooleanFieldProp.Value;
+
+        var jsonStr = JConvert.SerializeObject((ExpandoObject)expandoValue);
+        Assert.Equal("{\"stringValue\":\"test\",\"numberValue\":123,\"booleanValue\":true}", jsonStr);
```
```suggestion
       // Arrange
        var contentItem = new ContentItem();
        contentItem.Alter<TestPart>(part =>
        {
            part.TextFieldProp = new TextField { Text = "test" };
            part.NumericFieldProp = new NumericField { Value = 123 };
            part.BooleanFieldProp = new BooleanField { Value = true };
        });

        // Act
        dynamic expandoValue = new ExpandoObject();
        expandoValue.stringValue = contentItem.Content.TestPart.TextFieldProp.Text;
        expandoValue.numberValue = contentItem.Content.TestPart.NumericFieldProp.Value;
        expandoValue.booleanValue = contentItem.Content.TestPart.BooleanFieldProp.Value;

       // Assert
        var jsonStr = JConvert.SerializeObject((ExpandoObject)expandoValue);
        Assert.Equal("{\"stringValue\":\"test\",\"numberValue\":123,\"booleanValue\":true}", jsonStr);
```

hishamco commented on test/OrchardCore.Tests/Data/JsonDynamicTests.cs:
```diff
@@ -337,4 +341,479 @@ public void JsonDynamicValueMustConvertToUri()
+        dynamic myDynamic = new JsonDynamicValue(JsonValue.Create(expectedValue));
+
+        Assert.True(expectedValue == myDynamic);
+        Assert.False(expectedValue != myDynamic);
```
I don't think we need the second check while it's already checked before

gvkries replied:
No, both operators `==, !=` are implemented on `JsonDynamicValue` and should be tested. But I agree that these tests are not the most beautiful.

hishamco replied:
I see, so ignore my above comment

hyzx86 commented:
A week has passed. Nothing for now, is it mandatory that we use Lambada? 🤨

gvkries commented:
I don't see the remaining formatting suggestions as mandatory... 🙈

hyzx86 commented:
Hi @hishamco  ，All right, all right, let's merge it quickly.  
![f81e195dbb51b505036cba00ecfb23a0](https://github.com/OrchardCMS/OrchardCore/assets/15613121/10054b66-27e6-4430-ba01-1ea53a179189)

giannik commented:
merge, merge...

MikeAlhayek commented:
@gvkries I enabled auto merge. Hopefully you are done making changes :)

sebastienros commented:
> @sebastienros if there's any objection regarding expression bodies please let me, because it was a useful feature for single-line methods

I have stated my "personal" preferences and motivations several times already.
